File: ackermann.c
/*
 Ackermann's function in c.
 Faster than the one I wrote in python and it doesn't crash
 after calculating ackermann of 4, 1
 */

#include <stdio.h>

// does all the calculations
int ack(int m, int n) {
	int ans;
	
	if (m == 0) ans = n+1;
	else if (n == 0) ans = ack(m-1,1);
	else ans = ack(m-1, ack(m,n-1));
	return (ans);
}

// increments i and j and passes them as args to ack()
int main (int argc, char ** argv) { 
	int i,j;

	for (i=0; i<6; i++)
	for (j=0;j<6; j++)

	printf ("ackerman (%d,%d) is: %d\n",i,j, ack(i,j));
}

File: ackermann.py
'''
Ackermann's fuction in python

NOTE: after calculating ackermann of 4, 1
it crashes because python says: RecursionError: maximum recursion depth exceeded in comparison
'''

# this does all the calculations
def ack(m, n):
    
    m = m
    n = n
    ans = 0
    
    # the interesting bit
    if (m == 0):
        ans = n + 1
    elif (n == 0):
        ans = ack(m - 1, 1)
    else:
        ans = ack(m - 1, ack(m, n - 1))
    return ans

# increments i and j and passes them as args to ack()
def main():
    for i in range(6):
        for j in range(6):
            print("ackermann of (%d, %d) = %d" % (i, j, ack(i, j)))
            
if __name__ == '__main__':
    main()
    
